### PR TITLE
Heartbeat skip serialize and deserialize

### PR DESCRIPTION
**Pull request by zhangyz-hd**
## What is the purpose of the change

提供者和消费者之间的心跳，Request和Response的Data都为null：

```
//org.apache.dubbo.remoting.exchange.support.header.HeartbeatTimerTask#doTask
Request req = new Request();
req.setVersion(Version.getProtocolVersion());
req.setTwoWay(true);
req.setEvent(HEARTBEAT_EVENT);//HEARTBEAT_EVENT = null
```

```
//org.apache.dubbo.remoting.exchange.support.header.HeartbeatHandler#received
Response res = new Response(req.getId(), req.getVersion());
res.setEvent(HEARTBEAT_EVENT);//HEARTBEAT_EVENT = null
```

当提供者和消费者数量非常大时，每个节点都需要发送/接受大量的心跳Requet和Response，这些Requet、Response也会和服务调用一样经过serialize+encode、decode+deserialize，而其中的serialize/deserialize需要消耗过多的CPU资源。
这个优化的目的是对于心跳Requet/Response跳过serialize/deserialize处理，serialize改为赋值为null对象的序列化后字节，deserialize改为直接赋值null。

## Brief changelog

在CodecSupport中增加getNullBytesOf(Serialization s)方法，首次访问时根据serialization获得null对象的字节数组并缓存到ID_NULLBYTES_MAP
DubboCodec、ExchangeCodec中在调整encodeRequest、encodeResponse、decodeBody逻辑，实现上述目标。

备注：目前DUBBO的Serialization的实现中，以下三个实现其实尚不支持null对象的序列化，在CodecSupport.getNullBytesOf中降级为返回空字符串。

1. KryoSerialization2
2. GenericProtobufSerialization
3. GenericProtobufJsonSerialization

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/CodecSupport.java [hunk 1]
```diff
@@ -41,6 +44,8 @@
     private static Map<Byte, Serialization> ID_SERIALIZATION_MAP = new HashMap<Byte, Serialization>();
     private static Map<Byte, String> ID_SERIALIZATIONNAME_MAP = new HashMap<Byte, String>();
     private static Map<String, Byte> SERIALIZATIONNAME_ID_MAP = new HashMap<String, Byte>();
+    // Cache null object serialize results, for heartbeat request/response serialize use.
+    private static Map<Byte, byte[]> ID_NULLBYTES_MAP = new HashMap<Byte, byte[]>();
```
Comment (AlbumenJ) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/CodecSupport.java:48]: hi, @zhangyz-hd , will this map causes memory leak

Comment (zhangyz-hd) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/CodecSupport.java:48]: There will be no memory leaks,  it just cache the bytes of each Serialization-SPI null object when used. For example, '4e' for "hessian2", and '00' for 'kryo'.

Comment (AlbumenJ) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/transport/CodecSupport.java:48]: thanks for your explanation

**Review by AlbumenJ — APPROVED** [review]
LGTM.

**Review by chickenlj — APPROVED** [review]

**Pull request merged**